test/release/examples/primers/interopWithC.chpl

```diff
@@ -0,0 +1,332 @@
+
+   Compiling C Code with the Chapel Library
+   ++++++++++++++++++++++++++++++++++++++++
+*/
+
+/*
+   Compiling C code with the generated Chapel library file is generally complex,
+   but can be made simpler through the use of the ``--library-makefile`` flag
+   as described in :ref:`readme-libraries-linking`.
+   
+   An example of compiling a C program with a generated Chapel library using the
+   generated Makefile can be found in the `Makefile
+   <https://github.com/chapel-lang/chapel/blob/master/test/release/examples/primers/Makefile>`_
+   for the primers directory, to build this source file.
+*/ 
+
+/*
+   .. _primers-C-interop-using-C:
+
+   Using C Code in Chapel
+   ----------------------
+*/
+
+/*
+   Chapel has support for C code but we need to tell the compiler
+   about the C symbols using the ``extern`` keyword
+*/
+
+   extern proc baz(): int;
+
+//
+// The function can then be called as normal:
+//
+   proc callsBaz() {
+      writeln(baz());
+   }
+
+//
+// Such functions can even be called in exported functions:
+//
+
+   export proc alsoCallsBaz() {
+      writeln(baz());
+      callsBaz();
+   }
+
+// You can tell the chapel compiler where to look for these C functions by adding a
+// require statement
+   require "cHelper.h", "cHelper.c";
+// You must inlcude both a header file and C file.
+// Chapel also supports require statements for ``.o`` files and
+// for archived libraries using the ``-l`` flag
+/*
+   .. code-block:: chapel
+
+         require "foo.o", "-lapache-arrow";
+*/
+
+// Alternatively you can also include their names while invoking the chapel compiler.
+
+//
+// Unlike ``export``, ``extern`` can also be applied to global variables,
+// struct definitions, or even typedefs.
+//
+   extern var x: int(32);
+   extern var y: uint(32) = 3;
+   extern type myType = int;
+   extern proc useMyType(arg: myType): int; // an extern function using the typedef
+
+   export proc callUseMyType() {
+      var blah: myType = 17;
+      writeln(useMyType(blah));
+   }
+
+// Chapel also has a standard module named :mod:`SysCTypes` (located under
+//``$CHPL_HOME/modules/standard/gen/...``) which defines a few C types which
+// align with the C compiler specification and do not require the ``extern`` keyword, 
+// such as ``c_int`` and ``c_char``. For more information about these types see
+// the :ref:`readme-extern` Technical Note.
+
+// You can include SysCTypes using a simple use statement
+
+   use SysCTypes;
+
+// You can also assign default arguments to extern procs which then can be omitted
+// at the callsite as usual. For example
+
+   extern proc sum(a: c_int, b: c_int = 1): int;
+
+   writeln(sum(10, 10));
+   writeln(sum(10));
+// We must always make sure the types align as the C compiler specification allows for
+// different sizes for the same type depending on the compiler.
+
+// In order to pass an array to C, we use the following declaration:
+   extern proc sumArray(arr: [] int, size: c_int): c_int;
```

Comment:
Since it looks like you can pass Chapel arrays to C I guess I'd wonder what the purpose of c_array is.

Reply:
I am not sure of the differentiation myself. Maybe `c_array` was introduced before we were able to pass Chapel Arrays to C?
I can't think of why someone would still use `c_array` since Chapel Arrays are so much more powerful, unless there is a significant difference between them.
Maybe @lydia-duncan can shed some light on this.